### ISSUE TITLE
SNOW-1331246: Hashlib type hints breaks snowpark for python 3.9

### ISSUE THREAD

**Issue by tmcclintock** [snowflakedb/snowpark-python#1388]
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

```
Python 3.9.17 (main, Jun 29 2023, 09:32:35) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]
```

2. What operating system and processor architecture are you using?

`macOS-14.4.1-arm64-arm-64bit`

3. What are the component versions in the environment (`pip freeze`)?

<details>
  <summary>freeze</summary>
  
```txt
agate==1.7.1
annotated-types==0.6.0
appdirs==1.4.4
appnope==0.1.4
asn1crypto==1.5.1
asttokens==2.4.1
attrs==23.2.0
Babel==2.14.0
certifi==2024.2.2
cffi==1.16.0
chardet==5.2.0
charset-normalizer==3.3.2
click==8.1.7
cloudpickle==2.2.1
colorama==0.4.6
comm==0.2.2
cryptography==42.0.5
dbt-core==1.7.11
dbt-extractor==0.5.1
dbt-hubspot-dependencies==0.6.5
dbt-semantic-interfaces==0.4.4
dbt-snowflake==1.7.3
debugpy==1.8.1
decorator==5.1.1
diff_cover==8.0.3
exceptiongroup==1.2.0
executing==2.0.1
filelock==3.13.3
hypothesis==6.100.1
idna==3.6
importlib-metadata==6.11.0
iniconfig==2.0.0
ipykernel==6.29.4
ipython==8.18.1
isodate==0.6.1
jaraco.classes==3.4.0
jedi==0.19.1
Jinja2==3.1.3
jinja2-simple-tags==0.6.1
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
jupyter_client==8.6.1
jupyter_core==5.7.2
keyring==24.3.1
leather==0.4.0
Logbook==1.5.3
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mashumaro==3.11
matplotlib-inline==0.1.6
mdurl==0.1.1
minimal-snowplow-tracker==0.0.2
more-itertools==10.2.0
msgpack==1.0.8
multimethod==1.10
mypy-extensions==1.0.0
nest-asyncio==1.6.0
networkx==3.2.1
numpy==1.26.4
packaging==24.0
pandas==2.2.2
pandera==0.18.3
parsedatetime==2.6
parso==0.8.4
pathspec==0.11.2
pexpect==4.9.0
platformdirs==3.11.0
pluggy==1.4.0
prompt-toolkit==3.0.43
protobuf==4.25.3
psutil==5.9.8
ptyprocess==0.7.0
pure-eval==0.2.2
pycparser==2.22
pydantic==2.6.4
pydantic_core==2.16.3
Pygments==2.17.2
PyJWT==2.8.0
pyOpenSSL==24.1.0
pytest==8.1.1
python-dateutil==2.9.0.post0
python-slugify==8.0.4
pytimeparse==1.1.8
pytz==2024.1
PyYAML==6.0.1
pyzmq==25.1.2
referencing==0.34.0
regex==2023.12.25
requests==2.31.0
rich==13.7.1
rpds-py==0.18.0
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
six==1.16.0
snowflake-connector-python==3.7.1
snowflake-snowpark-python==1.14.0
sortedcontainers==2.4.0
sqlfluff==2.3.5
sqlfluff-templater-dbt==2.3.5
sqlparse==0.4.4
stack-data==0.6.3
tblib==3.0.0
text-unidecode==1.3
toml==0.10.2
tomli==2.0.1
tomlkit==0.12.3
tornado==6.4
tqdm==4.66.2
traitlets==5.14.2
typeguard==4.2.1
typing-inspect==0.9.0
typing_extensions==4.10.0
tzdata==2024.1
urllib3==1.26.18
wcwidth==0.2.13
wrapt==1.16.0
```
</details>

4. What did you do?

Recently installed the latest snowflake.snowpark, and now a basic import cannot run. All I ran is `import snowflake.snowpark`

<details>
  <summary>stacktrace</summary>
  
```
  File "/Users/tmcclintock/projects/scratch/failing_import.py", line 1, in <module>
    import snowflake.snowpark
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/__init__.py", line 47, in <module>
    from snowflake.snowpark.async_job import AsyncJob
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/async_job.py", line 12, in <module>
    from snowflake.snowpark._internal.analyzer.analyzer_utils import result_scan_statement
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/_internal/analyzer/analyzer_utils.py", line 10, in <module>
    from snowflake.snowpark._internal.analyzer.binary_plan_node import (
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/_internal/analyzer/binary_plan_node.py", line 7, in <module>
    from snowflake.snowpark._internal.analyzer.expression import Expression
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/_internal/analyzer/expression.py", line 9, in <module>
    import snowflake.snowpark._internal.utils
  File "/Users/tmcclintock/.pyenv/versions/3.9.17/envs/dbt/lib/python3.9/site-packages/snowflake/snowpark/_internal/utils.py", line 386, in <module>
    hash_algo: hashlib._hashlib.HASH, path: str, chunk_size: int, whole_file_hash: bool
AttributeError: module 'hashlib' has no attribute '_hashlib'
```
</details>

5. What did you expect to see?

I expected to be able to import anything from snowpark.

The stacktrace suggests that the type hint is busted and comes from referencing a private attribute of [`hashlib`](https://docs.python.org/3/library/hashlib.html). IMO, this is poor practice, because the `hashlib` maintainers clearly did not expect folks to be referencing private attributes. Can we please decouple snowpark from the internal API of `hashlib`? I suggest using a `Protocol` class or perhaps an enum that maps string names to hash algos. Something like:

```python
from enum import Enum

hash_algo = Enum('hash_algo', ['md5', 'sha512', 'sha256'])
```

6. Can you set logging to DEBUG and collect the logs?

The logging statement did not generate any additional output.

Thank you!

**Comment by sfc-gh-aling** [comment]
hi @tmcclintock , thanks for the feedback. weirdly I'm not able to reproduce the issue in my env. I checked the code in hashlib, so there is a chance that `_hashlib` can be None, and it might be caused by some dependency diff.

but I agree that type hint referencing a private module could be unreliable for, and would better be avoided or at least should not cause any error.

I have made a quick fix for it just to unblock: https://github.com/snowflakedb/snowpark-python/pull/1406/files.
